miner/agent.go

```diff
@@ -58,8 +61,12 @@ func (self *CpuAgent) Stop() {
 }
 
 func (self *CpuAgent) Start() {
-	self.mu.Lock()
 	defer self.mu.Unlock()
+	self.mu.Lock()
+
+	if atomic.LoadInt32(&self.isMining) == 1 {
+		return // agent already started
+	}
```

Comment:
Although in this specific instance this code is valid, it is only so because the atomic operations are protected by the mutex. The correct way to check whether it's running and to set it to 1 if it is not is:

```go
if !atomic.CompareAndSwapUint32(&self.isMining, 0, 1) {
        return
}
```